File: C++/next_perm.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    int index = -1;
    for(int i=n-2; i>=0; i--){
        if(nums[i] < nums[i+1]){
            index = i;
            break;
        }
    }
    if(index == -1){
        reverse(nums.begin(), nums.end());
    } else {
        for(int i=n-1; i>index; i--){
            if(nums[i] > nums[index]){
                swap(nums[i], nums[index]);
                break;
            }
        }
        reverse(nums.begin() + index + 1, nums.end());
    }
    for (int i = 0; i < n; i++) {
        cout << nums[i] << " ";
    }
    return 0;
}

File: Python/next_perm.py
n = int(input())
nums = []
for i in input().split():
    nums.append(int(i))

index = -1
for i in range(n-2, -1, -1):
    if nums[i] < nums[i+1]:
        index = i
        break

if index == -1:
    nums.sort(reverse=True)

else:
    for i in range(n-1, index, -1):
        if nums[i] > nums[index]:
            nums[i], nums[index] = nums[index], nums[i]
            break
    nums[index+1:n] = nums[index+1:n][::-1]
print(nums)